book/src/dev/rfcs/XXXX-asynchronous-script-verification.md

```diff
@@ -0,0 +1,231 @@
+
+This RFC describes an architecture for asynchronous script verification and
+its interaction with the state layer. This architecture imposes constraints
+on the ordering of operations in the state layer.
+
+# Motivation
+[motivation]: #motivation
+
+As in the rest of Zebra, we want to express our work as a collection of
+work-items with explicit dependencies, then execute these items concurrently
+and in parallel on a thread pool.  
+
+# Definitions
+[definitions]: #definitions
+
+- *UTXO*: unspent transaction output. Transaction outputs are modeled in `zebra-chain` by the [`TransparentOutput`][transout] structure.
+- Transaction input: an output of a previous transaction consumed by a later transaction (the one it is an input to).  Modeled in `zebra-chain` by the [`TransparentInput`][transin] structure.
+- lock script: the script that defines the conditions under which some UTXO can be spent.  Stored in the [`TransparentOutput::lock_script`][lock_script] field.
+- unlock script: a script satisfying the conditions of the lock script, allowing a UTXO to be spent.  Stored in the [`TransparentInput::PrevOut::lock_script`][lock_script] field.
+
+[transout]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.TransparentOutput.html
+[lock_script]: https://doc.zebra.zfnd.org/zebra_chain/transaction/struct.TransparentOutput.html#structfield.lock_script
+[transin]: https://doc.zebra.zfnd.org/zebra_chain/transaction/enum.TransparentInput.html
+[unlock_script]: https://doc.zebra.zfnd.org/zebra_chain/transaction/enum.TransparentInput.html#variant.PrevOut.field.unlock_script
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Zcash's transparent address system is inherited from Bitcoin. Transactions
+spend unspent transaction outputs (UTXOs) from previous transactions. These
+UTXOs are encumbered by *locking scripts* that define the conditions under
+which they can be spent, e.g., requiring a signature from a certain key.
+Transactions wishing to spend UTXOs supply an *unlocking script* that should
+satisfy the conditions of the locking script for each input they wish to
+spend.
+
+This means that script verification requires access to data about previous
+UTXOs, in order to determine the conditions under which those UTXOs can be
+spent. In Zebra, we aim to run operations asychronously and out-of-order to
+the greatest extent possible. For instance, we may begin verification of a
+block before all of its ancestors have been verified or even downloaded. So
+we need to design a mechanism that allows script verification to declare its
+data dependencies and execute as soon as all required data is available.
+
+It's not necessary for this mechanism to ensure that the transaction outputs
+remain unspent, only to give enough information to perform script
+verification. Checking that all transaction inputs are actually unspent is
+done later, at the point that its containing block is committed to the chain.
```

Comment:
This is a really important design pattern, which we might end up re-using for other data dependencies.